dwave/system/temperatures.py

```diff
@@ -0,0 +1,488 @@
+                    replace=True)
+                T_bootstrap_estimates[bs],_ = maximum_pseudolikelihood_temperature(
+                    site_energy = site_energy[indices],
+                    bootstrap_size = 0,
+                    Tguess = T_estimate)
+    
+    return T_estimate, T_bootstrap_estimates
+
+def freezeout_effective_temperature(freezeout_B,temperature,units_B = 'GHz',units_T = 'mK'):
+    ''' Provides an effective temperature as function of freezeout information.
+    
+    A D-Wave quantum annealer is assumed to implement a Hamiltonian
+    H = B(s)/2 H_P - A(s)/2 H_D. H_P is the problem (classical) Hamiltonian
+    and H_D is the driver Hamiltonian. B(s) is the problem energy scale and A(s)
+    is the driver (transverse field) energy scale, s is the normalized anneal
+    time s = t/t_a (in [0,1]).
+
+    If a quantum annealer achieves an equilibrated distribution 
+    over decohered states late in the anneal then the transverse field is 
+    negligible: A(s) << B(s). 
+    If in addition dynamics stop abruptly the equilibrated distribution is 
+    described by a classical Boltzmann distribution for classical spin states s 
+    that may be measured in accordance with a probability distribution:
+    P(s) = exp(- B(s*) H_P(s) / 2 kB T)
+    B(s*) is the schedule energy scale associated to the problem Hamiltonian, 
+    T is the physical temperature and kB is the Boltzmann constant.
+    We can define a unitless effective temperature, to complement the unitless
+    Hamiltonian, definition as T_eff = 2 kB T/B(s*), returned by this function.
+    P(s) = exp(-H_P(s)/T_eff)
+
+    Single qubit freeze-out (s*) is well characterized as part of calibration 
+    processes and reported alongside annealing schedules {A(s),B(s)} and 
+    device temperature. This allows an effective temperature to be calculated 
+    for online solvers, appropriate for some simple Hamiltonians. Values are
+    typically specified in mK and GHz. More complicated systems such as chains
+    may freeze-out differently (at different s, or asynchronously). Large
+    problems may have slow dynamics at small s, where A(s) cannot be ignored.
+    One should also be aware of known noise model properties for purposes 
+    of interpretting a temperature.
+
+    Note that for QPU solvers this temperature applies to programmed 
+    Hamiltonians H_P submitted with auto_scale = False. If auto_scale=True 
+    (default) an additional scaling factor is required.
+
+    Args:
+        freezeout_B (float):
+            The schedule value for the problem Hamiltonian at freeze-out.
+
+        temperature (float):
+            The physical temperature of the annealer.
+        
+        units_B (string, optional, 'GHz'):
+            Units in which the schedule is specified. Allowed values:
+            'GHz' (Giga-Hertz) and 'J' (Joules).
+
+        units_T (string, optional, 'mK'):
+            Units in which the schedule is specified. Allowed values:
```

Comment:
"the schedule" --> either "the temperature" or better still, "in which ``temperature`` is specified"